frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -19,112 +19,106 @@ limitations under the License.
+            if (!rv->empty()) break; } }
+    return rv;
+}
+
+std::vector<const IR::IDeclaration*>*
+ResolutionContext::lookup(const IR::INamespace *current, IR::ID name, P4::ResolutionType type,
+                          bool forwardOK) const {
+    LOG2("Trying to resolve in " << current->toString());
+
+    if (current->is<IR::IGeneralNamespace>()) {
+        auto gen = current->to<IR::IGeneralNamespace>();
+        Util::Enumerator<const IR::IDeclaration*> *decls = gen->getDeclsByName(name);
+        switch (type) {
+            case P4::ResolutionType::Any:
+                break;
+            case P4::ResolutionType::Type: {
+                std::function<bool(const IR::IDeclaration*)> kindFilter =
+                        [](const IR::IDeclaration *d) {
+                    return d->is<IR::Type>(); };
+                decls = decls->where(kindFilter);
+                break; }
+            case P4::ResolutionType::TypeVariable: {
+                std::function<bool(const IR::IDeclaration*)> kindFilter =
+                        [](const IR::IDeclaration *d) {
+                return d->is<IR::Type_Var>(); };
+                decls = decls->where(kindFilter);
+                break; }
+        default:
+            BUG("Unexpected enumeration value %1%", static_cast<int>(type)); }
+
+        if (!forwardOK) {
+            std::function<bool(const IR::IDeclaration*)> locationFilter =
+                    [name](const IR::IDeclaration *d) {
                 Util::SourceInfo nsi = name.srcInfo;
-                Util::SourceInfo dsi = decl->getNode()->srcInfo;
+                Util::SourceInfo dsi = d->getNode()->srcInfo;
                 bool before = dsi <= nsi;
                 LOG2("\tPosition test:" << dsi << "<=" << nsi << "=" << before);
-                if (!before)
-                    continue;
-            }
+                return before; };
+            decls = decls->where(locationFilter); }
 
+        auto vector = decls->toVector();
+        if (!vector->empty()) {
             LOG2("Resolved in " << dbp(current->getNode()));
-            auto result = new std::vector<const IR::IDeclaration*>();
-            result->push_back(decl);
-            return result;
-        }
+            return vector; }
+    } else {
+        auto simple = current->to<IR::ISimpleNamespace>();
+        auto decl = simple->getDeclByName(name);
+        if (decl == nullptr)
+            return &empty;
+        switch (type) {
+            case P4::ResolutionType::Any:
+                break;
+            case P4::ResolutionType::Type: {
+                if (!decl->is<IR::Type>())
+                    return &empty;
+                break; }
+            case P4::ResolutionType::TypeVariable: {
+                if (!decl->is<IR::Type_Var>())
+                    return &empty;
+                break; }
+        default:
+            BUG("Unexpected enumeration value %1%", static_cast<int>(type)); }
+
+        if (!forwardOK) {
```

Comment:
Is this code duplicated?

Reply:
This is the old code (mostly) unchanged, but the indentation changed (moved it out of the loop into a separate `lookup` function, so that it could be called on just the global scope for absolute paths, rather than creating a new ResolveContext with just the global scope on its lookup stack).

I'm thinking of refactoring this by making much of it a method on ISimpleNamespace/IGeneralNamespace, using vitual dispatch rather than dynamic cast tests.